src/paymentservice/src/main/java/ca/bc/gov/open/pssg/rsbc/dps/paymentservice/BamboraClientImpl.java

```diff
@@ -0,0 +1,184 @@
+package ca.bc.gov.open.pssg.rsbc.dps.paymentservice;
+
+import java.net.URL;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.UUID;
+
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.lang3.StringUtils;
+
+import ca.bc.gov.open.pssg.rsbc.dps.paymentservice.PaymentServiceConstants.BamboraTransType;
+import ca.bc.gov.open.pssg.rsbc.dps.paymentservice.exception.PaymentServiceException;
+
+/**
+ * Bambora Client class Implementation
+ * 
+ * May be expanded in the future to include other types of Bambora transactions. 
+ * 
+ * @author smillar
+ *
+ */
+public class BamboraClientImpl implements PaymentClient {
+
+	private URL hostedPaymentURL = null;
+	private String merchantId = null;
+	private String hashKey = null;
+	private int minutesToExpire;
+	private String correlationId = null;
+	
+	public URL getHostedPaymentServiceURL() {
+		return hostedPaymentURL;
+	}
+
+	public String getMerchantId() {
+		return merchantId;
+	}
+
+	public URL getHostedPaymentURL() {
+		return hostedPaymentURL;
+	}
+
+	public int getMinutesToExpire() {
+		return minutesToExpire;
+	}
+	
+	public String getCorrelationId() {
+		return correlationId;
+	}
+
+	// constructor
+	public BamboraClientImpl(URL hostedPaymentURL, String merchantId, String hashKey, int minutesToExpire) throws PaymentServiceException {
+		this.hostedPaymentURL = hostedPaymentURL;
+		this.merchantId = merchantId;
+		this.hashKey = hashKey;
+		this.minutesToExpire = minutesToExpire;
+	}
+
+	/**
+	 * 
+	 * Calculate Single Payment URL 
+	 * 
+	 * @param correlationId
+	 * @param transType
+	 * @param invoiceNumber
+	 * @param totalItemsAmount
+	 * @param totalGST
+	 * @param totalPST
+	 * @param approvedPage
+	 * @param declinedPage
+	 * @param errorPage
+	 * @param ref1
+	 * @param ref2
+	 * @param ref3
+	 * @return
+	 * @throws PaymentServiceException
+	 */
+	@Override
+	public URL calculateSinglePaymentURL(String correlationId, BamboraTransType transType, String invoiceNumber, double totalItemsAmount, double totalGST, 
+			double totalPST, String approvedPage, String declinedPage, String errorPage, String ref1, String ref2, String ref3)
+			throws PaymentServiceException {
+		
+		// add correlationId if non provided in request. 
+		if (StringUtils.isEmpty(correlationId)) {
+			this.correlationId = generateUniqueCorrelationId();
+	    }
+		
+		String redirect = null;
+		
+		try {
+	
+			// Add the optional parameters if available.
+			String errorParam = (errorPage != null) ? "&" + PaymentServiceConstants.BAMBORA_PARAM_ERROR_PAGE + "=" + errorPage
```

Comment:
awesome use of ternary operator here and below. super clean. =)

Reply:
👍